File: packages/go/pkg/bot/inline.go
package bot

import (
	"context"

	"github.com/Nam088/telebot/packages/go/pkg/types"
)

// AnswerInlineQuery sends an answer to an inline query.
//
// Returns true on success, or an error if the API call fails.
//
// Telegram API: https://core.telegram.org/bots/api#answerinlinequery
func (b *Bot) AnswerInlineQuery(ctx context.Context, opts *types.AnswerInlineQueryOptions) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "answerInlineQuery", opts, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// AnswerWebAppQuery sets the result of an interaction with a Web App.
//
// Returns a SentWebAppMessage on success, or an error if the API call fails.
//
// Telegram API: https://core.telegram.org/bots/api#answerwebappquery
func (b *Bot) AnswerWebAppQuery(ctx context.Context, opts *types.AnswerWebAppQueryOptions) (*types.SentWebAppMessage, error) {
	var msg types.SentWebAppMessage
	if err := b.Request(ctx, "answerWebAppQuery", opts, &msg); err != nil {
		return nil, err
	}
	return &msg, nil
}

// SavePreparedInlineMessage stores a message that can be sent by the user of a Mini App.
//
// Returns a PreparedInlineMessage on success, or an error if the API call fails.
//
// Telegram API: https://core.telegram.org/bots/api#savepreparedinlinemessage
func (b *Bot) SavePreparedInlineMessage(ctx context.Context, opts *types.SavePreparedInlineMessageOptions) (*types.PreparedInlineMessage, error) {
	var msg types.PreparedInlineMessage
	if err := b.Request(ctx, "savePreparedInlineMessage", opts, &msg); err != nil {
		return nil, err
	}
	return &msg, nil
}

// SavePreparedKeyboardButton saves a prepared keyboard button for a Mini App.
//
// Returns the PreparedKeyboardButton the docs declare as this method's result,
// which carries the identifier callers reuse to attach the button, or an error
// if the API call fails.
//
// Example:
//
//	button, err := b.SavePreparedKeyboardButton(ctx, &types.SavePreparedKeyboardButtonOptions{
//		UserID: int64(123456),
//		Button: map[string]any{"text": "Pay", "pay_for_access": true},
//	})
//
// Telegram API: https://core.telegram.org/bots/api#savepreparedkeyboardbutton
func (b *Bot) SavePreparedKeyboardButton(ctx context.Context, opts *types.SavePreparedKeyboardButtonOptions) (*types.PreparedKeyboardButton, error) {
	var button types.PreparedKeyboardButton
	if err := b.Request(ctx, "savePreparedKeyboardButton", opts, &button); err != nil {
		return nil, err
	}
	return &button, nil
}

File: packages/python/src/telebot_py/bot/inline.py
"""Inline-mode Bot API methods (parity with packages/go/pkg/bot/inline.go)."""

from __future__ import annotations

from collections.abc import Sequence

from telebot_py.bot.base import (
    MarkupLike,
    Requester,
    clean_payload,
    parse_flag,
    parse_result,
    to_wire,
)
from telebot_py.types.message_extras import PreparedInlineMessage, SentWebAppMessage


class InlineMixin(Requester):
    """Bot methods for answering inline queries from search results."""

    async def answer_inline_query(
        self,
        inline_query_id: str,
        results: Sequence[MarkupLike],
        *,
        cache_time: int | None = None,
        is_personal: bool | None = None,
        next_offset: str | None = None,
        button: MarkupLike | None = None,
    ) -> bool:
        """Send answers to an inline query.

        Example:
            >>> ok = await bot.answer_inline_query(query.id, [result.to_dict()])

        Args:
            inline_query_id: Unique identifier of the answered query.
            results: Inline query results; dicts or ``to_dict`` objects.
            cache_time: Seconds the result may be cached on the server.
            is_personal: Pass True if results may be cached only for the
                querying user; by default results are public.
            next_offset: Offset the client should use in the next query.
            button: Inline query results button (InlineQueryResultsButton).

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.

        Telegram API: https://core.telegram.org/bots/api#answerinlinequery
        """
        payload = clean_payload(
            inline_query_id=inline_query_id,
            results=[to_wire(result) for result in results],
            cache_time=cache_time,
            is_personal=is_personal,
            next_offset=next_offset,
            button=to_wire(button),
        )
        return parse_flag(await self.request("answerInlineQuery", payload))

    async def answer_web_app_query(
        self, web_app_query_id: str, result: MarkupLike
    ) -> SentWebAppMessage:
        """Set the result of an interaction with a Web App.

        Example:
            >>> message = await bot.answer_web_app_query(query_id, result)

        Args:
            web_app_query_id: Unique identifier of the Web App query.
            result: The result description as an InlineQueryResult dict.

        Returns:
            A SentWebAppMessage with the sent message identifier.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.

        Telegram API: https://core.telegram.org/bots/api#answerwebappquery
        """
        payload = clean_payload(
            web_app_query_id=web_app_query_id,
            result=to_wire(result),
        )
        return parse_result(SentWebAppMessage, await self.request("answerWebAppQuery", payload))

    async def save_prepared_inline_message(
        self,
        user_id: int,
        result: MarkupLike,
        *,
        allow_user_chats: bool | None = None,
        allow_bot_chats: bool | None = None,
        allow_group_chats: bool | None = None,
        allow_channel_chats: bool | None = None,
    ) -> PreparedInlineMessage:
        """Store a message a Mini App user can send to a chat later.

        Example:
            >>> prepared = await bot.save_prepared_inline_message(42, result)

        Args:
            user_id: Unique identifier of the user who may send the message.
            result: The message description as an InlineQueryResult dict.
            allow_user_chats: Allow sending to private chats with users.
            allow_bot_chats: Allow sending to private chats with bots.
            allow_group_chats: Allow sending to group and supergroup chats.
            allow_channel_chats: Allow sending to channel chats.

        Returns:
            The stored PreparedInlineMessage.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.

        Telegram API: https://core.telegram.org/bots/api#savepreparedinlinemessage
        """
        payload = clean_payload(
            user_id=user_id,
            result=to_wire(result),
            allow_user_chats=allow_user_chats,
            allow_bot_chats=allow_bot_chats,
            allow_group_chats=allow_group_chats,
            allow_channel_chats=allow_channel_chats,
        )
        return parse_result(
            PreparedInlineMessage,
            await self.request("savePreparedInlineMessage", payload),
        )
